### ISSUE TITLE
requirements: write spec for DIDComm libraries

### ISSUE THREAD

**Issue by dhh1128** [sicpa-dlab/didcomm-gemini#5]
We need a design doc that describes the requirements and architecture approach for a library that "implements" the DIDComm v2 spec. (Obviously the spec itself is the main artifact we could compare a library to, to see if it implements the right stuff. The question is: *How should features of DIDComm be bundled into callable functions and data structures such that they are conveniently usable in a given programming language?* That's what this ticket is asking for.)

**Comment by dhh1128** [comment]
Here is a first cut at a proposal: https://hackmd.io/@dhh1128/r11pkclT_

**Comment by CluEleSsUK** [comment]
I'm getting a 404 for this unfort: 
![image](https://user-images.githubusercontent.com/3749956/124557553-ef0c6080-de39-11eb-8d27-739c6eedf06f.png)

**Comment by dhh1128** [comment]
See if this URI works better:

https://hackmd.io/@dhh1128/r11pkclT_/edit

**Comment by dhh1128** [comment]
Two more sections needed:

1. Something about types of crypto supported.
2. Something about error handling.

**Comment by vimmerru** [comment]
@dhh1128 I think we are missing quite important abstraction here. DID Resolver doesn't help to access private keys of KeyAgreement. So we need something like `Wallet` and important topic is choosing between crypto enclave or direct access to private keys.

**Comment by dhh1128** [comment]
I suggest that we deliberately keep Wallet out of scope. I agree that something needs to manage keys -- but I am looking for code more like SSH code that uses keys without asking where you got them, so I want to make that something a problem for the consumers of the library, not for the libraries themselves. The wallet abstraction introduced a *huge* amount of complexity into Indy SDK, and also introduced quite a lot of controversy (many community members had strong opinions about how the wallet we provided didn't meet their needs).

Can we simply require that private keys be passed as arguments to functions, and not attempt to solve the lookkup problem?

(If that's truly not possible, then I guess we could create a private key lookup interface that doesn't have any impls in our library, and require a reference to that interface in our functions. That's as far as I'd want to go.)

**Comment by vimmerru** [comment]
> Can we simply require that private keys be passed as arguments to functions, and not attempt to solve the lookkup problem?

There is a potential problem with this approach. What if new DID Specification will requite a new crypto? It is quite probable scenario keeping in ming crypto was changed 2 times recent 6 months. As i understand we just want to make a something similar to "openssl". A super crypt library with a lot of legacy crypto to support and a lot of crypto negotiation-related vulnerabilities. Now DIDComm spec says nothing about this, but what if ECDH-1PU v4 based library want to communicate with ECDH-1PU v3 based library written 2 months ago?

**Comment by dhh1128** [comment]
I agree that your ECDH-1PU example is important. This is particularly true because ECDH-1PU is just a draft, and it has changed once already; the chances that it will change again before it is finished are reasonably high.

But I don't understand why this leads to the conclusion that we need a wallet abstraction. Maybe it's because I am expecting the wallet to do something different than you are? Does a wallet have to do the actually crypto? Or does it just have to look up the private key that corresponds to the public key?

**Comment by dhh1128** [comment]
Here is ACAPY and its relationship to wallets, today:

![acapy today](https://i.ibb.co/jJnyKRC/ACAPY-wallets-today.png)

One way we could fit our libraries into ACAPY's architecture and resolve the tension between private key passing and the design of wallets in Indy is like this:

![option 1](https://i.ibb.co/GtJS2MH/ACAPY-wallets-option-1.png)

**Comment by dhh1128** [comment]
And here is a more radical option, just for comparison. In this option, the current wallet construct is decomposed, and maintenance of security for private keys becomes an app-level concern:

![option 2](https://i.ibb.co/h1VMcGn/ACAPY-wallets-disassembled-option-2.png)

I am not advocating option 2. I am just showing some different perspectives on how to solve the problem. What I want to assert is that:

1. There are ways to do this that involve only modest refactors to ACAPY
2. We don't need a full blown wallet abstraction in our didcomm libs. It is too "heavy."

What ideas do others have?

**Comment by ashcherbakov** [comment]
@dhh1128 
A couple of questions:
1) Do we understand https://github.com/decentralized-identity/universal-resolver by DID universal resolver?
2) The lib spec says that it's sufficient to support Ed25519 and secp256k1 curves only, but DID Comm spec mentions support for ES256 (NIST p-256) for signing (https://identity.foundation/didcomm-messaging/spec/#algorithms) and P-384 and P-256 for encryption (https://identity.foundation/didcomm-messaging/spec/#curves-and-content-encryption-algorithms)
3) Do I understand correctly, that examples of `encrypt` and `decrypt` methods are not exact function signatures, and rather just say what we want from the requirements point of view? I believe the real methods may look a bit different as
 - need to distinguish authcrypt and anoncrypt
 - need to specify the algorithm for encryption and algorithm for key wrapping 
 - the current methods in the spec use DIDs as inputs, which means that DID resolving will be done under the hood. That can be fine in some cases, but if DID resolution requires complex network or asynchronous communication, it may be better to separate the resolution and the crypto.

**Comment by dhh1128** [comment]
#1: yes
#2: Perhaps it would be best to just start with Ed25519 support and then circle back to add support for extra curves later. Having support for the first curve (Edwards) is 70% of the value of the library. I added secp256k1 because I thought that would make the library attractive to Ethereum folks, which is a constituency that feels important to me. However, I may have judged wrong and it's actually ES256 that's the next most important. This is especially likely if the DIDComm spec lists it, because other people have already had a thoughtful conversation about what to support. Nobody is arguing that Edwards isn't first, so that is safe. 
#3: yes -- those are just rough examples. I agree with your observations about ways this needs to change.

**Comment by ashcherbakov** [comment]
4. Do we have any specific requirements about platforms / OS for the libs? It seems a requirement that the libs must be run on Linux, Windows, Macos, Android, iOS, doesn't it? We may want to make sure that it's true when working on CI/CD.

**Comment by dhh1128** [comment]
>Do we have any specific requirements about platforms / OS for the libs

Thank you for asking this question. It's an obvious one, in hindsight.

Yes. Libraries should run on every platform where the language in question is mainstream. I don't know that python is very important on mobile, and I wouldn't want to spend a lot of effort making sure our python lib runs there. But JavaScript/TS is mainstream on mobile, so it should run there. Etc. Please make a concrete proposal about CI/CD details, and we can ratify it as a group.

**Comment by dbluhm** [comment]
> And here is a more radical option, just for comparison. In this option, the current wallet construct is decomposed, and maintenance of security for private keys becomes an app-level concern. DIDComm libraries written in Gemini are still plugged in, but now it's at the app level rather than inside the wallet:
> 
> ![option 2](https://camo.githubusercontent.com/9917be2f4d337b4edc0da0bfcafa27c381c1fbdf0391341ec5ed7ad814ddf120/68747470733a2f2f692e6962622e636f2f6831564d63476e2f41434150592d77616c6c6574732d646973617373656d626c65642d6f7074696f6e2d322e706e67)
> 
> I am not advocating option 2. I am just showing some different perspectives on how to solve the problem. What I want to assert is that:
> 
> 1. There are ways to do this that involve only modest refactors to ACAPY
> 2. We don't need a full blown wallet abstraction in our didcomm libs. It is too "heavy."
> 
> What ideas do others have?

As it turns out, this is not actually too radical an option. ACA-Py is already structured to operate this way when using Askar. I agree with your assertion that we don't need a full blown wallet abstraction in the DIDComm libs and, given my findings on Askar, that this should only require modest refactors to ACA-Py. I'm continuing to scope things out on ACA-Py integration and will report any limitations I discover that might impact the interface we expose with the DIDComm libs. As of now, though, I'm feeling pretty confident that where the library spec stands now will not present any insurmountable obstacles.

**Comment by ashcherbakov** [comment]
> But JavaScript/TS is mainstream on mobile, so it should run there. Etc.

I believe we may discuss it in more details. Running Kotlin and JS for mobile is probably not a problem for peerdid lib as it doesn't have any crypto dependencies. But DID Comm libs have them. So, let's try to understand what languages should support what mobile platforms:

1) Android
- Java/Kotlin for sure 
- JS: there can be problems with react native depending what crypto/JOSE libs we will use. For example, https://github.com/panva/jose#runtime-support-matrix doesn't work on React Native.
2) iOS
- Kotlin multiplatform - needs to research
- JS: same as for Android
- Do we need a native Swift implementation? 

We've been evaluating some of this open questions during design and JOSE evaluation phase, but probably some input from the requirements point of view could also be useful.
For example, we may say from the requirements point of view, that we are going to support only Native mobile libs (Kotlin for Android and Swift for iOS), and JS will target node.js and browser (probably limited support) only.

**Comment by dhh1128** [comment]
The general principle we are trying to respect, regarding platform support, is: **Get the most bang for the buck.** We don't need anything to run in an environment where it is unnatural or extremely expensive to build. Instead, we want to create obvious features -- things that most of the developers on a platform will ask for. So we don't have to do Kotlin on iOS, for example, if we believe it's a marginal value with a high expense. The overarching goal is to aid adoption, not to provide something that's cross-platform.

**Comment by ashcherbakov** [comment]
OK, makes sense. 

What about the following **proposal for the first release**: let's target the following platforms for the first release:
- Python: Windows/Linux/Mac
- Java/Kotlin: Windows/Linux/Mac/Android
- JS: Windows/Linux/Mac, nodes.js (backend) + limited support in browsers (25519 curve may not work, https://www.w3.org/TR/WebCryptoAPI/)

This proposal doesn't require iOS support. If iOS support is must have for the first release, then probably we should consider ObjectiveC/Swift native library? It can naturally be used by native iOS developers, as well as by ReactNative developers (with a wrappers around native Kotlin/Swift libs).
Also we may investigate (as the second phase after the first phase is released), whether and how hard it would be to support iOS via Kotlin multiplatform or JS/ReactNative.

**Comment by dhh1128** [comment]
I think browser support ought to be relatively inexpensive, because I think [TweetNaclJS](https://www.npmjs.com/package/tweetnacl) demonstrates that Ed25519 support is available in browser JS. @ashcherbakov , you linked the WebCryptoAPI stuff; does TweetNaclJS make you more optimistic? I would really not like to give up support in browsers.

I would also like to have *something* that targets iOS. I would be willing to give up some other parts of the platform matrix (e.g., python on mac, Kotlin on Mac) to get it. But I am guessing that the iOS support is going to be too expensive because it means Swift, and that's a whole new language. Maybe I'll just have to live without it. But maybe we could put it in the backlog and see whether we have time in September?

**Comment by vimmerru** [comment]
@dhh1128 @ashcherbakov TweetNaclJS seems native JS port of TweetNaCl. So it isn't based on WebCryptoAPI and implements all primitives on native javascript.

Most probably it can't be used as basis for JOSE as TweetNaCl exposes only high-level crypto primitives. Technically we can expose them in the fork, but trust to such crypto will be quite low.

**Comment by dhh1128** [comment]
@vimmerru : Can you say more about "trust to such crypto will be quite low"? Are you imagining that trust to WebCryptoAPI will be higher? If so, how much higher, and why?

**Comment by ashcherbakov** [comment]
As for iOS support, we think that the best way to provide adaption of DIDComm for mobile platforms is to implement native libs. We're already targeting Android as part of Kotlin. But as for iOS, a native ObjectiveC/Swift would probably be the best choice, so that it can be adapted by any Mobile framework that targets iOS.
So, my proposal is the following: let's add tickets for native iOS support to the backlog. It will be the next priority after the basic Python/Kotlin/JS support. 

As for TweetNaCl, what Vyacheslav means, is that TweetNaCl doesn't implement ECDH-1PU. It has high level authenticated encryption based on  x25519-xsalsa20-poly1305 (same way as authcrypt worked in Indy). But in order to implement ECDH-1PU, we will need low level primitives (curves, operations, crypo, etc.). TweetNaCl doesn't expose these primitives. So, if we want to base our ECDH-1PU implementation on TweetNaCl. one of the options would be to fork TweetNaCl and expose the low primitives there. And as we depend on not the original TweetNaCl, but on our fork (with changed API), people may have lower trust to our implementation and our fork.

**Comment by vimmerru** [comment]
>  And as we depend on not the original TweetNaCl, but on our fork (with changed API), people may have lower trust to our implementation and our fork.

Even more the main idea of TweetNaCl/NaCl is provide opinionated high level API that will allow people to avoid building of custom crypto system as low-level crypto misusage is the main attack vector. 

What we can do is copy paste low level functions from TweetNaCl and combine them to API that will allow us to use crypto we need, but it will be our "own" crypto library without any adoptions or peer reviews. Trust to such libraries is quite low in crypto-community.

**Comment by dhh1128** [comment]
Thank you for the clarifying comments, @ashcherbakov and @vimmerru .

>let's add tickets for native iOS support to the backlog. It will be the next priority after the basic Python/Kotlin/JS support.

Agreed.

**Comment by dhh1128** [comment]
I am content to close this issue. Do you agree, @ashcherbakov and @vimmerru ?

**Comment by ashcherbakov** [comment]
@dhh1128 @vimmerru 
Probably we have just one question left related to requirements (realized it when started working on the API).

If we follow "explicit" API Option, so that all keys (private, public) are passed directly to API methods of our lib (as it's suggested in the discussions for this ticket), then it looks like we need neither Wallet nor DID resolving interfaces.
Example of API:
```
bytes_to_transfer = MessageBuiler(payload)
    .from(from_did)
    .build()
    .sign(
        from=(from_sign_key_id, from_sign_private_key)
    )
    .auth_crypt(
        alg=ECDH-1PU+A256KW,
        enc=A256CBC-HS512,   
        from=(from_key_id, from_private_key),
        to=[
            (to_key_id_1, to_public_key_1),
            (to_key_id_2, to_public_key_2),
        ]
    )
```
Here it's assumed that all private keys are obtained on the calling App level from an App-specific wallet, and all public keys (and key IDs) are obtained from a resolved DID DOC. DID resolving and DID DOC parsing looks out of scope for common DID Comm methods in this option, and can be done on the App level by calling an appropriate DID resolver and DID parser.

However, the requirements say, that DID resolving interface must be part of DIDComm libs. Is it still a requirement if we go with "explicit API" option as above? Is it assumed that DID DOC parsing be part of DIDComm lib, or it can/should be done by other libraries?

**Comment by dhh1128** [comment]
I believe that the explicit API option still requires DID resolution to be part of the DIDComm libs, because there are calculations based on DID state that must be done right in the middle of the crypto. (I argued against this, as hard as I could, when DIDComm was being designed. My claim was that we must keep the crypto and the DIDs entirely separate. But I was basically the only person in the WG who took that position, in the end. Everybody else insisted that it was no big deal to do DID resolution right in the middle of a crypto operation. So I don't like this characteristic of DIDComm, but I believe we can't avoid it.)

Even if the above comment is not true, I think the API shown above is too "pure" in its approach to the "explicit" ideal. When Alice sends to Bob, it is reasonable to require her to identify the specific key she wants to use as the sender; there is no way for a library to figure that out. But it is counter to the design of DIDComm for Alice to have to enumerate all of Bob's keys. The mental paradigm should be that she is sending to Bob's DID. So I would like the API to take as in put Alice's sender key (or signer key), but NOT Bob's recipient keys. I want the API to look up Bob's recipient keys via a DID resolution interface (where the impl of that interface is not provided by the DIDComm lib).

**Comment by dhh1128** [comment]
Tagging @TelegramSam ; please comment if you like, Sam.

**Comment by dhh1128** [comment]
Cross-referencing my comments about explicit APIs [here](https://github.com/sicpa-dlab/didcomm-gemini/issues/18#issuecomment-884156587)

**Comment by ashcherbakov** [comment]
@dhh1128 
Do you agree that it will make sense to separate resolving of DIDs and DID Comm crypto operations (as in Option 2 https://hackmd.io/7POHiPtqQW6p5bQLRfAdXA?view#Option-2-Explicit-API), because we would like to keep crypto and potential network communication separate? I think this is the main thing we need to decide.
- If it's not the case, then we can just go with Option 1 which has very high level API like `encrypt(from_did, to_dids[]`).
- If we think that this separation is needed for everyone, we should go with Option 2. And I think that the properties you mentioned can be achieved by Option 2 even without DID resolving and DID parser interfaces embedded there. It may be not so clear from the API as-is, but can be more clear when look at the whole workflow (as shown in the example https://hackmd.io/7POHiPtqQW6p5bQLRfAdXA?view#Option-2-Explicit-API)
   - Alice doesn't need to enumerate Bob's keys explicitly. The enumeration will be done automatically: Alice calls DID resolver and DID DOC parsers, that will return all keys for the given Bob's DID. And then all these keys (as-is) are put to the `encrypt` method
   - As for checking if the key used in keyAgreement for encryption really belongs to Alice's expected DID, that can be achieved by passing the expected Alice's DID to `parse_encrypted(expected_sender_did)` method. The method cam check that `skid` matches the expected sender's DID and throw an exception otherwise. If it matches, we can cal DID resolver and parser (out-of-scope for DID Comm lib) and pass obtained keys (as-is) to the `decrypt` method. 
- If we think that this separation is needed, but not for everyone, we can go with Option 4, where we support two types of the API.
 

If so, then I have a feeling that although DID resolving and DID DOC parsing are essential parts of the workflow, we can make them out of scope, as in Option 2.

**Comment by dhh1128** [comment]
A part of me wants to say "yes, go with Option 2." But I don't believe it will be ideal. One reason lies in the assumption inside this paragraph you wrote:

```
Alice doesn't need to enumerate Bob's keys explicitly. The enumeration will be done automatically: Alice calls DID resolver and DID DOC parsers, that will return all keys for the given Bob's DID. And then all these keys (as-is) are put to the encrypt method
```

Everything you say in this paragraph is accurate. But it assumes that users of DIDComm libraries will *know* to call the DID resolver and DID Doc parsers, get a certain chunk of data out of them, and pass that chunk of data unmodified to the explicit APIs. Where is that knowledge going to be conveyed -- in DIDComm library docs, or in DID resolution library docs, or somewhere else?

I don't believe this is a realistic assumption. It means there will be a lot of tribal knowledge required of consumers of a DIDComm library. I would rather have a DIDComm library that knows that this lookup must be done, and that its results must be passed without modification, to the crypto -- even though it doesn't know what library is providing the lookup feature, or how that library works internally. That way users of the DIDComm libraries have less to learn, and less opportunities to do the wrong thing.

The other reason I am uncomfortable with the "explicit or not explicit" distinction you are trying to draw is that it treats DID and key info from a remote party the same way that it treats key info from the local caller of the library. I don't think that's correct, either. When I send to someone else, *I am thinking in terms of recipient DIDs, not in terms of recipient keys*. But as a sender, I am forced to think about keys, even if I have a wallet, because I have to decide which key I'm using. And if the roles are reversed (I'm the recipient, and the remote party is the sender), I should only be interested in sender keys to confirm correctness; it's mostly the sender's DID that interests me.) So trying to make the APIs purely explicit, both for the recipient side and the sender side, or purely high level, both for the recipient side and the sender side, feels wrong. They should be explicit for the local side (eliminating the need for a wallet interface), but high-level for the remote side (requiring DID resolution on an as-needed basis, but mostly allowing the programmer to think of, describe, and log the remote side by DID).

I think DIDComm libraries should mostly focus on the "Comm" and the crypto. But I think they should know *what* DID info about the remote party they need to configure the crypto correctly, rather than asking the caller to have that knowledge. This means the DIDComm libs should have an interface to DID resolution, and should call it.

**Comment by dhh1128** [comment]
As I re-read my comment above, I am left with a feeling of confusion. That's not a good sign. It means I'm not even convinced by my own argument. :-)

Maybe we need some other people to weigh in here...

**Comment by CluEleSsUK** [comment]
>they should know what DID info about the remote party they need to configure the crypto correctly, rather than asking the caller to have that knowledge. This means the DIDComm libs should have an interface to DID resolution, and should call it.


What I imagine from your suggestion is that the library having its own representation of a DID document containing only the relevant bits, and callers implementing an adapter from a separate DIDDocument parsing library of their choice to the DIDComm format, e.g.


```
// inside the didcomm lib
type DIDCommKey = ArrayBuffer

type DIDDocSubset = {
    key: DIDCommKey
}

interface DIDResolver {
    fetch(didUri: string): Promise<DIDDocSubset>
}
```

```
// inside some other DID doc parsing library
type DIDDocument = {
    endpoints: Array<Endpoint>
    methods: Array<methods>
    // ... etc - the full DID document
}

function parse(bytes: ArrayBuffer): DIDDocument {}
```


```
// implementation in my code

function didcommAdapter(did: DIDDocument): DIDDocSubset {
    // extracts just the relevant bits
}

class MyDIDResolver implements DIDResolver {
    fetch(didUri: string): Promise<DIDDocSubset> {
        // fetch from local file
        let didDoc = parse(someBytesFromFile)
        return didcommAdapter(didDoc)
    }
}

class UniversalResolver implements DIDResolver {
    fetch(didUri: string): Promise<DIDDocSubset> {
        // fetches the doc using the universal resolver
        let didDoc = parse(someBytesFromTheUniversalResolver)
        return didcommAdapter(didDoc)
    }
}

```

I'm being overly concrete here, but it might help guide our thoughts.

**Comment by ashcherbakov** [comment]
@dhh1128 
After more thinking, and listening to various arguments, we are actually proposing to go with Option 1 (implicit API): [Proposed Options](https://hackmd.io/7POHiPtqQW6p5bQLRfAdXA?view#Proposed-Option). Please look at the arguments there and let's discuss them.

Also we added a [Questions](https://hackmd.io/7POHiPtqQW6p5bQLRfAdXA?view#Questions) sections. The main question that may affect option choice is how onion (nested) encryption is expected to work.

Also the following changes have been done to the Options doc (https://hackmd.io/7POHiPtqQW6p5bQLRfAdXA):
- Note1 and Note2  have been added to [DID-Comm-workflow-components](https://hackmd.io/7POHiPtqQW6p5bQLRfAdXA?view#DID-Comm-workflow-components)
- Wallet interface has been renamed to Secrets Resolver interface 
- [Proposed Options](https://hackmd.io/7POHiPtqQW6p5bQLRfAdXA?view#Proposed-Option) section has been added
- A couple of cons and pros have been added to the Options (see [Options-Cons-and-Pros](https://hackmd.io/7POHiPtqQW6p5bQLRfAdXA?view#Options-Cons-and-Pros). In particular, related to the onion (nested) encryption
- [Option 2](https://hackmd.io/7POHiPtqQW6p5bQLRfAdXA?view#Option-2-Explicit-API) and [Option 3](https://hackmd.io/7POHiPtqQW6p5bQLRfAdXA?view#Option-3-Explicit-API-with-DID-Resolver-interface) examples have been extended with an onion (nested) encryption.
- [Questions](https://hackmd.io/7POHiPtqQW6p5bQLRfAdXA?view#Questions) have been added

**Comment by dhh1128** [comment]
I am finding it difficult to become confident that I understand your thinking because of an ambiguity in the following sentence from your options doc (in the description of option 1):

>All four components are part of DID Comm libs

Do you mean by this that the DIDComm libs contain all *implementation* for all 4 components -- or do you mean that the DIDComm libs contain *interface definitions plus calls to those interfaces* for secrets and for DID resolution, plus *implementation* for the crypto and communication logic?

If you mean the first one, then I am not aligned. If you mean the second one, then I am happy with option 1.

**Comment by ashcherbakov** [comment]
@dhh1128 
I mean `B`, only interfaces and calls to these interfaces for DID resolving and secrets, no implementation.

Actually we might add some default implementations for test purposes, but I believe it doesn't change the essence of the option.

As you can see from [DID-Comm-workflow-components](https://hackmd.io/7POHiPtqQW6p5bQLRfAdXA?view#DID-Comm-workflow-components), it talks about interfaces for DID resolving and secrets, not implementations. We were discussing potential implementation for DID Doc parser, but I believe we are on the same page now that DID Comm should have only the interface for DID DOC parser (probably as part of DID resolving interface).

**Comment by ashcherbakov** [comment]
Current design proposal we have a consensus about: https://hackmd.io/aD4TzKRhT1ma2ERl6EKU_g

- We defined allowed combinations for nested messages: [Possible-nested-message-combinations](https://hackmd.io/aD4TzKRhT1ma2ERl6EKU_g?view#Possible-nested-message-combinations)
- Receiver will need to call just a single `unpack` regardless of a message combination type
  - It will process and decrypt all nested parts including `anoncrypt(authcrypt)`
  - Receiver needs to be able to specify options for unpacker related to [MTC](https://github.com/hyperledger/aries-rfcs/blob/master/concepts/0029-message-trust-contexts/README.md) (as flags or using a builder pattern).
  - There should be options/flags for unpack to support unpack by mediators (anoncrypt(plain) is expected)
- `unpack` returns the payload, signed (JWS) message (can be empty), and metadata to be processed against MTC latter.
- It's OK to have a `forward` method that will take a payload (end-user-encrypted message) and prepare an onion of anoncrypted forward messages for the given DID
- We defined what consistency checks should be performed by the lib: [Proposed-Consistency-checks](https://hackmd.io/aD4TzKRhT1ma2ERl6EKU_g#Proposed-Consistency-checks)
- We are OK to have either asynchronous or synchronous version of the API depending on the language. We may provide both - synchronous and asynchronous versions where it makes sense if it doesn't delay our release date.
- We are addressing a possible issue with the proposed API that it can be too heavy (possible network communication etc.) under the hood by saying that it depends on the implementation of resolvers (DID, secrets). If one wants a simple and fast execution of API calls, he or she can always provide fast and simple implementations for resolvers.
- The API will allow to provide resolvers as arguments to the main calls keeping a possibility to register default resolvers.

**Comment by ashcherbakov** [comment]
I believe we discussed all major things related to requirements and API, so I'm fine to close the ticket.
@vimmerru @TelegramSam ?

**Comment by vimmerru** [comment]
@ashcherbakov I agree

**Comment by TelegramSam** [comment]
I believe there is a conflict in audience that I have not seen discussed that is very relevant:

Library Implementers: Want to see how the nuts and bolts work at a basic level. They want to see where keys are passed, not just identifiers.

Library Users: Want to see things at a higher level of abstraction. They care about identifiers.

I think we can solve this by developing the code in two layers: high level and 'low' level.

pack and unpack deal with DIDs, and are configured with a resolver and secret interfaces.

underlying methods like encrypt, decrypt, etc. use keys and DIDs in parallel. 

Both layers are available for calling and use and are part of the same library. Documentation directs the audience appropriately.

I believe this will not be a huge adjustment to the design, but will better serve the needed purposes moving forward. Thoughts? @dhh1128 @ashcherbakov @vimmerru

**Comment by vimmerru** [comment]
> Both layers are available for calling and use and are part of the same library. Documentation directs the audience appropriately.

In general i am against of exposing low level API as it just contradicts best practices for crypto libraries. Also it will require safe design for such api and dedicated unit/integration tests. I would expect cost of implementation will be increased to about 30%.

**Comment by ashcherbakov** [comment]
I agree with Vyacheslav.
We discussed the options for implicit (DID-based) and explicit (keys-based) APIs, and came to the conclusion that explicit API will require too much knowledge about DID Comm protocol from the user. Also we don't see any real cases where a person would need an explicit version instead of implicit one. 
So, we should ask the question: are there any real cases where developers will need low-level explicit APIs? 
The next thing we need to answer: if we do see such a case for a low-level explicit API, is it a real need for a low-level API, or the DID Comm spec lacks some important details, so a developer has to find a workaround?
Even if there are such cases, I believe 95% of DID Comm users will need and use implicit (DID-based) API. So, my proposal is to start with implicit (high-level) API only. If we or community realize that a low-level version is also needed, it can be added in future releases. There should be no technical issues for this.

**Comment by TelegramSam** [comment]
I could be convinced not to expose the lower-level layers, but distinctly representing them inside the code will be a HUGE help to those learning the internals of the code, including those porting to other languages.

**Comment by ashcherbakov** [comment]
Sure, the low-level calls will be represented in the code. High-level calls (exposed API) will consist of a number of smaller calls (low-level APIs) in the appropriate order.

**Comment by ashcherbakov** [comment]
@TelegramSam @dhh1128 
Are we ready to close the issue? Or there is anything else to discuss?

**Comment by dhh1128** [comment]
Yes, ready to close.

**Comment by dhh1128** [comment]
Although I closed this issue, I want to leave an additional comment.

I agree with Sam that the low-level calls could be quite helpful, so I was encouraged when I read Alex's comment:

>Sure, the low-level calls will be represented in the code. High-level calls (exposed API) will consist of a number of smaller calls (low-level APIs) in the appropriate order.

Both the high level and the low level will of course be functions. But we could do a couple things to give Sam some of the value he's imagining, for relatively low effort:

* In languages that have visibility restrictions (public vs. private), we might consider making some of the lower-level functions public.
* The main high-level APIs will be documented -- not by writing separate docs, but by providing very helpful doc comments that include detailed descriptions of preconditions, parameters, errors, and return values. We could consider adding comments like these to a few of the lower-level functions, as well.

Now, I make these suggestions knowing that it could be easy to go too far. We don't want to expose all the guts of a library, and we don't want to help people shoot themselves in the foot by trying to combine primitives in a way that doesn't make sense. So the test of whether it's actually worth doing either of these two suggestions (and if yes, how *much* it's worth doing them) boils down to Alex's question:

>are there any real cases where developers will need low-level explicit APIs?

I predict that the answer to these questions will be "no" for the next few weeks, but will gradually become "yes" as we learn exactly how the libraries will be used. I think we're going to get in a situation where we want to route a message through TOR, or where we want to do something advanced with attachments, or where we want to deliberately send a message to less than all of Alice's keys, and we'll realize that it's worth decomposing the opionated, optimized top-level abstractions just a little bit.

So my process suggestion is to be vigilant -- ready to notice AND CAPTURE evidence that the answer to Alex's question is "yes". As we do, we should be okay with the idea that some advanced usage ought to be possible. This doesn't mean the API is undermined in its goal to make things easy and to help programmers do the right thing. It just means that there are some corner cases that are real. And when that happens, we expose/document the lower-level functions AS NEEDED. The benefit is that programmers don't become frustrated by our libraries as they evolve from DIDComm beginners to DIDComm experts.
